feat/equalize3d

vribeiro1 opened the pull request:
### Description
Implements equalize function for a sequence of images using PyTorch ops. Considers the histogram of the volume on the equalization calculations.

_Note for the reviewer: The original code for equalize function does not have tests._

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated


## PR Checklist
### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

shijianjian commented:
LGTM. Can you add ```kornia.augmentation.augmentation3d.RandomEqualize3D``` as well?

I only wrote the doctest for ```augmentation.py``` back to that time, please feel free to add new test. For running doctest, you may use ```python -m doctest kornia/augmentation/augmentation.py```

edgarriba reviewed (changes requested):
unit test for new functionalities are mandatory too.

edgarriba commented:
@shijianjian @vribeiro1 open a separate PR for `RandomEqualize3D`.
@shijianjian regarding doctest, could you add an automatic trigger for it ?

vribeiro1 commented:
@shijianjian @edgarriba I will write the appropriate tests and open a second PR for RandomEqualize3D. Thank you for your comments!

shijianjian reviewed (changes requested):
LGTM. Please add more test.

shijianjian commented on test/enhance/test_adjust.py:
```diff
@@ -599,3 +599,150 @@ def test_gradcheck(self, device):
+        assert_allclose(f(inputs), expected)
+
+    def test_equalize_batch(self, device):
+        bs, channels, depth, height, width = 2, 3, 6, 20, 20
```
```suggestion
        bs, channels, height, width = 2, 3, 20, 20
```

shijianjian commented on test/enhance/test_adjust.py:
```diff
@@ -599,3 +599,150 @@ def test_gradcheck(self, device):
+        assert f(inputs).shape == torch.Size([bs, channels, height, width])
+
+    def test_equalize(self, device):
+        bs, channels, depth, height, width = 1, 3, 6, 20, 20
```
```suggestion
        bs, channels, height, width = 1, 3, 20, 20
```

edgarriba commented on kornia/enhance/adjust.py:
```diff
@@ -430,46 +462,35 @@ def equalize(input: torch.Tensor) -> torch.Tensor:
     return torch.stack(res)
 
 
+def equalize3d(input: torch.Tensor):
```
missing return type in definition `def equalize3d(input: torch.Tensor) -> torch.Tensor:`

edgarriba commented on kornia/enhance/adjust.py:
```diff
@@ -430,46 +462,35 @@ def equalize(input: torch.Tensor) -> torch.Tensor:
+        input (torch.Tensor): image tensor with shapes like (C, D, H, W) or (B, C, D, H, W) to equalize.
+    Returns:
+        torch.Tensor: Sharpened image or images.
+    """
```
docs won't render well - add this instead:
r"""Equalizes the values for a 3D volumetric tensor.

Implements Equalize function for a sequence of images using PyTorch ops based on uint8 format:
    https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/autoaugment.py#L352

Args:
        input (torch.Tensor): image tensor with shapes like :math:`(C, D, H, W)` or :math:`(B, C, D, H, W)` to equalize.

Returns:
        torch.Tensor: Sharpened image or images with same shape as the input.
"""

edgarriba commented on kornia/enhance/adjust.py:
```diff
@@ -430,46 +462,35 @@ def equalize(input: torch.Tensor) -> torch.Tensor:
+    Returns:
+        torch.Tensor: Sharpened image or images.
+    """
+    input = _transform_input3d(input) * 255.
```
shouldn't we scale the tensor in first place inside `_scale_channel` ?

edgarriba commented on test/enhance/test_adjust.py:
```diff
@@ -599,3 +599,150 @@ def test_gradcheck(self, device):
+
+
+class TestEqualize:
+    def test_shape_equalize(self, device):
```
add also `dtype` to tests `def test_shape_equalize(self, device, dtype):`

edgarriba commented on test/enhance/test_adjust.py:
```diff
@@ -599,3 +599,150 @@ def test_gradcheck(self, device):
+    def test_shape_equalize(self, device):
+        bs, channels, height, width = 1, 3, 4, 5
+
+        inputs = torch.ones(channels, height, width)
```
pytorch recommendation to initialize tensors:
`inputs = torch.ones(channels, height, width, device=device, dtype=dtype)`

edgarriba commented on test/enhance/test_adjust.py:
```diff
@@ -599,3 +599,150 @@ def test_gradcheck(self, device):
+
+
+class TestEqualize3D:
+    def test_shape_equalize3d(self, device):
```
same as above for `dtype`

edgarriba reviewed (changes requested):
LGTM. Just few minor changes

vribeiro1 commented:
As requested by @shijianjian, I have developed the code for `RandomEqualize3D`. I will open a new PR on top of this branch so you can review. 
On this new PR, I added tests for `RandomEqualize` as well, and fixed a potential bug on it after testing.

shijianjian commented:
@vribeiro1 Can you rebase? Your commits are duplicated twice here.

vribeiro1 commented:
@shijianjian Sorry about that. I rebased and double checked. Now it looks fine. I also solved a conflict we had with master in one docstring. If you find anything weird, please, let me know :)

shijianjian commented on kornia/enhance/adjust.py:
```diff
@@ -428,60 +429,86 @@ def _blend_one(input1: torch.Tensor, input2: torch.Tensor, factor: Union[float,
+
+def equalize(input: torch.Tensor) -> torch.Tensor:
+    r"""Apply equalize on the input tensor.
+    
```
```suggestion
```
